### PR TITLE
feat: parse CycloneDX scope field for optional/excluded dep detection

### LINKED ISSUE #303: feat: use CycloneDX scope field to auto-detect runtime dependencies in diet analysis
## Summary

Currently, `uzomuzo diet` uses a hardcoded `mavenRuntimeDeps` whitelist (#300) to recognize reflection-loaded dependencies (JDBC drivers, logging backends, WebJars) and prevent false-positive UNUSED reports.

Investigation into the CycloneDX `scope` field revealed a **critical limitation**: the CycloneDX spec defines only three scope values (`required`, `optional`, `excluded`), and **both Maven compile and runtime scopes map to `required`**. Therefore, CycloneDX scope **cannot** replace the `mavenRuntimeDeps` whitelist for runtime detection.

However, the `scope` field provides value for two other categories:

| CycloneDX scope | Maven scope | Action |
|-----------------|-------------|--------|
| `required` | compile, runtime | No action (default). Runtime detection uses whitelist. |
| `optional` | provided | Annotate as `ScopeOptional`. Provided-scope deps (e.g., `javax.servlet-api`) are supplied by the runtime environment. They typically DO have source imports, so `IsUnused` is NOT suppressed. |
| `excluded` | test | Filter out from diet plan. Test deps should not appear in production analysis. |

## Proposed Changes

1. **Add `Scope` field to `sbomgraph.Component`** — parse CycloneDX scope from JSON
2. **Add `BuildScopeMap` utility** — walks components recursively, returns `map[normalizedPURL]scope`
3. **Add `ScopeByPURL` to `GraphResult`** — propagate scope through the graph analysis pipeline
4. **Filter `excluded` deps** — remove test-scope deps from diet plan (like `filterWorkspaceDeps`)
5. **Annotate `optional` deps** — set `ScopeOptional` so output explains provided-scope semantics
6. **Fallback to `mavenRuntimeDeps` whitelist** — when SBOM has no scope info (Trivy, syft)
7. **Update ADR-0015** — document the hybrid strategy and CycloneDX scope limitation

## Scope Availability by SBOM Tool

| Tool | Scope field | Notes |
|------|-------------|-------|
| cdxgen | Yes (`required`/`optional`/`excluded`) | Best coverage |
| CycloneDX Maven Plugin | Yes (same three values) | Native Maven scope mapping |
| Trivy | No | Whitelist fallback needed |
| syft | No | Whitelist fallback needed |

## Key Design Decision

`ScopeOptional` does NOT suppress `IsUnused` (unlike `ScopeTool`/`ScopeRuntime`). Provided-scope deps like `javax.servlet-api` typically have real static imports — the coupling analysis is accurate for them. The scope annotation is informational only.

## References

- #300: `ScopeRuntime` whitelist (current approach)
- #302: ADR-0015 Java reflection detection strategy (updated)
- #248: JDBC drivers flagged as unused (parent issue)

### PR DESCRIPTION
**Pull request by kotakanbe**
## Summary

- Parse CycloneDX `scope` field from SBOM components to detect provided-scope (`optional`) and test-scope (`excluded`) dependencies
- Filter `excluded` (test) deps from the diet plan — they are not production dependencies
- Annotate `optional` (provided) deps with `ScopeOptional` — informational label, does NOT suppress `IsUnused` since provided deps typically have real source imports
- Update ADR-0015 to document the hybrid strategy: SBOM scope for optional/excluded + `mavenRuntimeDeps` whitelist for runtime detection
- Document per-language diet analysis capabilities and Java-specific constraints in `docs/diet.md`

**Key finding**: CycloneDX scope cannot distinguish compile from runtime (both map to `required`), so the `mavenRuntimeDeps` whitelist (#300) remains essential for runtime-scope detection.

Closes #303

## Before / After

Synthetic CycloneDX SBOM (Spring Boot project with cdxgen-style scope fields): 9 deps total — 5 required (compile/runtime), 2 optional (provided: `servlet-api`, `lombok`), 2 excluded (test: `junit-jupiter`, `mockito-core`).

### Before (main)

Test deps appear in the diet plan as regular entries. Provided deps have no scope annotation.

```
── Diet Plan (9 direct dependencies) ─────────────────────────

RANK  SCORE  EFFORT   PURL                                                          REMOVES  IMPORTS  STATUS
────  ─────  ──────   ────                                                          ───────  ───────  ──────
1     0.10   trivial  pkg:maven/javax.servlet/javax.servlet-api@4.0.1               0        0        Archived
2     0.03   trivial  pkg:maven/org.projectlombok/lombok@1.18.30                    0        0        Active
3     0.02   trivial  pkg:maven/org.springframework/spring-web@6.1.0                0        0        Active
4     0.02   trivial  pkg:maven/org.mockito/mockito-core@5.11.0          ← test     0        0        Active
5     0.02   easy     pkg:maven/ch.qos.logback/logback-classic@1.5.6                0        0        Active
6     0.02   easy     pkg:maven/org.postgresql/postgresql@42.7.0                    0        0        Active
7     0.02   trivial  pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.17.0  0        0        Active
8     0.02   trivial  pkg:maven/com.google.guava/guava@33.0.0                       0        0        Active
9     0.02   trivial  pkg:maven/org.junit.jupiter/junit-jupiter@5.10.0   ← test     0        0        Active
```

JSON scope output:
```
Total direct: 9

  javax.servlet/javax.servlet-api        scope=(none)     ← no annotation
  org.projectlombok/lombok               scope=(none)     ← no annotation
  org.mockito/mockito-core               scope=(none)     ← test dep, no filtering
  ch.qos.logback/logback-classic         scope=runtime
  org.postgresql/postgresql              scope=runtime
  org.junit.jupiter/junit-jupiter        scope=(none)     ← test dep, no filtering
  ...
```

### After (this PR)

Test deps (`junit-jupiter`, `mockito-core`) are **filtered out** — they no longer appear in the diet plan. Provided deps (`servlet-api`, `lombok`) are **annotated** with `scope=optional`.

```
── Diet Plan (7 direct dependencies) ─────────────────────────

RANK  SCORE  EFFORT   PURL                                                          REMOVES  IMPORTS  STATUS
────  ─────  ──────   ────                                                          ───────  ───────  ──────
1     0.10   trivial  pkg:maven/javax.servlet/javax.servlet-api@4.0.1               0        0        Archived
2     0.03   trivial  pkg:maven/org.projectlombok/lombok@1.18.30                    0        0        Active
3     0.02   trivial  pkg:maven/org.springframework/spring-web@6.1.0                0        0        Active
4     0.02   easy     pkg:maven/ch.qos.logback/logback-classic@1.5.6                0        0        Active
5     0.02   easy     pkg:maven/org.postgresql/postgresql@42.7.0                    0        0        Active
6     0.02   trivial  pkg:maven/com.fasterxml.jackson.core/jackson-databind@2.17.0  0        0        Active
7     0.02   trivial  pkg:maven/com.google.guava/guava@33.0.0                       0        0        Active
                                                          ↑ junit-jupiter, mockito-core filtered out
```

JSON scope output:
```
Total direct: 7                                          ← was 9

  javax.servlet/javax.servlet-api        scope=optional   ← NEW: annotated as provided
  org.projectlombok/lombok               scope=optional   ← NEW: annotated as provided
  ch.qos.logback/logback-classic         scope=runtime
  org.postgresql/postgresql              scope=runtime
  ...
                                         ← junit-jupiter, mockito-core: gone (excluded)
```

### Key differences

| Aspect | Before | After |
|--------|--------|-------|
| Direct dep count | 9 | 7 (test deps removed) |
| Test deps (junit, mockito) | Appear as normal entries | Filtered out |
| Provided deps (servlet-api, lombok) | No scope annotation | `scope=optional` |
| Runtime deps (postgresql, logback) | `scope=runtime` (whitelist) | `scope=runtime` (unchanged) |
| SBOM without scope (Trivy, syft) | — | No change (whitelist fallback) |

## Changes

| Layer | File | Change |
|-------|------|--------|
| Infrastructure | `sbomgraph/types.go` | Add `Scope` field to `Component`, add `BuildScopeMap` |
| Infrastructure | `depgraph/graph.go` | Populate `ScopeByPURL` in `GraphResult` |
| Domain | `diet/types.go` | Add `ScopeOptional` constant, `ScopeByPURL` field |
| Application | `diet/service.go` | Add `filterExcludedDeps`, optional scope annotation |
| Interfaces | `cli/diet_render.go` | Render optional scope in detailed view |
| Docs | ADR-0015 | Amend with hybrid scope strategy |
| Docs | `docs/diet.md` | Java constraints + per-language capability tables |

## Test plan

- [x] `TestBuildScopeMap` — scope extraction from flat/nested components, nil/empty edge cases
- [x] `TestFilterExcludedDeps` — nil map, empty map, excluded filtering, optional/required preservation
- [x] `TestRun_OptionalScopeAnnotation` — optional deps get `ScopeOptional`, coupling data preserved
- [x] `TestRun_OptionalScopeNotOverriddenByRuntime` — `mavenRuntimeDeps` whitelist takes precedence
- [x] `TestRun_ExcludedDepsFilteredOut` — excluded deps removed from diet plan
- [x] All existing tests pass
- [x] CI: lint, test, CodeQL all pass

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by kotakanbe** [comment]
@claude Follow the procedure in `.github/prompts/review.prompt.md` with `--copilot-only` mode for this PR.

Phase 2: Resolve all unresolved Copilot review threads (discover → classify → fix → build/test/lint → commit/push → reply/resolve).
Phase 3: Detect recurring patterns from FIX-classified comments, propose and apply coding rules to `.github/instructions/`, then run `make sync-instructions`.
Phase 4: Post a summary comment on this PR.

**Comment by kotakanbe** [comment]
**Claude finished @kotakanbe's task in 3m 52s** —— [View job](https://github.com/future-architect/uzomuzo-oss/actions/runs/24327833446)

---
### Copilot Review Resolution (`--copilot-only`) — PR #304

- [x] Discover unresolved Copilot threads (2 found)
- [x] Classify each thread
- [x] Apply fixes, build/test/lint
- [x] Commit and push (`1ba209f`)
- [x] Reply and resolve all threads (2/2)
- [x] Phase 3: Pattern detection and rule promotion (`786b0a7`)
- [x] Phase 4: Summary

---

### Copilot Resolution — PR #304

| # | File | Classification | Action |
|---|------|---------------|--------|
| 1 | `internal/infrastructure/sbomgraph/types.go:338` | **FIX** | Added `slog.Warn` on max nesting depth in `buildScopeMapRecursive`, consistent with `buildRefMapRecursive` |
| 2 | `internal/application/diet/service.go:98` | **WONT_FIX** | Follows pre-existing `filterWorkspaceDeps` pattern — both filter post-AnalyzeGraph; refactoring the pipeline is out of scope |

Commits: `1ba209f`, `786b0a7`
Threads resolved: 2/2

### Rules Learned

| Pattern | Status | Detail |
|---------|--------|--------|
| defensive-coding | **Promoted** → `copilot-learned-coding.instructions.md` | 2 entries across PRs #281, #304 — "Mirror Sibling Function Behavior for Logging and Error Handling" |
 | [Branch](https://github.com/future-architect/uzomuzo-oss/tree/worktree-feat-303-sbom-scope)

File: internal/infrastructure/sbomgraph/types.go [hunk 1]
```diff
@@ -319,6 +320,37 @@ func flattenAggregatorModules(
...
+	buildScopeMapRecursive(components, m, 0)
+	if len(m) == 0 {
+		return nil
+	}
+	return m
+}
+
+// buildScopeMapRecursive populates m with normalizedPURL → CycloneDX scope mappings.
+func buildScopeMapRecursive(components []Component, m map[string]string, depth int) {
+	if depth > MaxNestingDepth {
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/infrastructure/sbomgraph/types.go:338]: Fixed in 1ba209f.

File: internal/application/diet/service.go [hunk 2]
```diff
@@ -86,10 +86,15 @@ func (s *Service) Run(ctx context.Context, input DietInput) (*domaindiet.DietPla
...
+	postWorkspaceCount := len(graphResult.DirectDeps)
+	graphResult.DirectDeps = filterExcludedDeps(graphResult.DirectDeps, graphResult.ScopeByPURL)
+
 	slog.Info("Phase 1 complete",
 		"direct", len(graphResult.DirectDeps),
 		"totalTransitive", graphResult.TotalTransitive,
-		"workspaceDepsFiltered", preFilterCount-len(graphResult.DirectDeps),
+		"workspaceDepsFiltered", preFilterCount-postWorkspaceCount,
+		"excludedDepsFiltered", postWorkspaceCount-len(graphResult.DirectDeps),
 	)
```
[the remark this answers is not in the document]
Comment (kotakanbe) [reply internal/application/diet/service.go:102]: Not applicable — this follows the same post-AnalyzeGraph filtering pattern as the pre-existing `filterWorkspaceDeps` (line 87). Both filters run after graph analysis. Fixing this would require a larger refactor of the graph analysis pipeline, which is out of scope for this PR.

**Comment by kotakanbe** [comment]
Superseded by #308 (docs only). Code changes (CycloneDX scope parsing) deferred — practical value is limited since Trivy already excludes test deps and the optional annotation is informational only. See #303 for the design if implementation is needed later.
